### ISSUE TITLE
Buf cannot run breaking changes detection against remote branches

### ISSUE THREAD

**Issue by menishmueli** [bufbuild/buf#376]
Running `buf breaking --against .git#branch=origin/main` is failing with the following error
```
warning: Could not find remote branch origin/main to clone.
fatal: Remote branch origin/main not found in upstream origin
```

Running `buf breaking --against .git#fef=origin/main` is failing with the following error
```
Failure: could not clone file:///Users/myyser/myrepo/.git: exit status 1
error: pathspec 'origin/main' did not match any file(s) known to git

```
Trying to walkaround this issue and running `git rev-parse origin/main` and then `buf breaking --against .git#ref={git_sha}` also sometimes cause the following error:
```
Failure: could not clone file:///home/myuser/dev/myrepo/.git: exit status 128
fatal: reference is not a tree: {git_sha}
unlinkat /tmp/{git_sha}: directory not empty.                                                                                                                                                                                         
```

`buf.yaml` config for breaking changes detection is standard. The root is the entire repo (default behavior) with some ignores, and the breaking rules are the default.

Running breaking changes detection against origin/main making a lot of sense instead of running against maybe old local main. This is true especially in CI, where you do not have local main branch when running on a feature branch before merging.

**Comment by manpreetszr** [comment]
Seems like a duplicate of https://github.com/bufbuild/buf/issues/339

**Comment by bufdev** [comment]
We've updated the git reference cloning change the way references and branches are treated. Previously, when specifying `.git#branch=origin/master`, Buf would `git clone` the repository at `$PWD/.git` and then checkout the ref `origin/master`, which would be confusing as `origin/master` referred to the `master` branch (after the clone `origin` would be `$PWD/.git`).

In the new behavior, instead of doing a clone we init a new repository and add `$PWD/.git` as a remote, after which we fetch the ref user supplied ref from the remote. This causes the ref to be interpreted in the context of the origin (`$PWD/.git`), and results in fetching the ref that makes more sense.

This should resolve this issue, let us know otherwise!

**Comment by bufdev** [comment]
This is released in 0.47.0.
